### DISCUSSION TITLE
Improve emcc/em++(Emscripten) support.

### DISCUSSION THREAD

**Discussion by lzralbu** [ccache/ccache#1472] [Support]
I can't figure out a way to make ccache(4.10) work along emcc(3.1.60) on Windows 10. I read somewhere that it isn't a officially supported compiler, but I wonder if it is supported at all, at least on Windows.

A simple "Hello world" in C won't compile. 

On command prompt:
```
c:\dev\emscripten-madness>emcc.bat main.c -o index.html
```

works alright, while
```
c:\dev\emscripten-madness>ccache emcc.bat main.c -o index.html
'C:\dev\tools\emsdk\upstream\emscripten\emcc.bat" "main.c" "-o" "index.html' is not recognized as an internal or external command,
operable program or batch file.
```  
don't.

**Comment by jrosdahl** [comment]
> ```
> c:\dev\emscripten-madness>ccache emcc.bat main.c -o index.html
> 'C:\dev\tools\emsdk\upstream\emscripten\emcc.bat" "main.c" "-o" "index.html' is not recognized as an internal or external command, operable program or batch file.
> ```

That output does not come from ccache, so something else is happening than you think (ccache is likely not executed at all). It seems like the _"is not recognized as an internal or external command, operable program or batch file"_ part is [printed by the Windows cmd.exe program](https://www.google.com/search?q=%22is+not+recognized+as+an+internal+or+external+command%2C+operable+program+or+batch+file.%22).

**Comment by lzralbu** [comment]
Indeed, there was something off. I made a small change to the command(changed from `emcc.bat` to `emcc`) and now I get the following output instead:

```
c:\dev\emscripten-madness>ccache emcc main.c -o index.html
ccache: error: execute_noreturn of C:\dev\tools\emsdk\upstream\emscripten\emcc failed: No such file or directory
```

I also tried masquerading emcc as someone suggested for a similar output in MSVC case but to no avail.

Any idea why?

**Reply by jrosdahl** [reply]
Does `C:\dev\tools\emsdk\upstream\emscripten\emcc` exist?

**Reply by lzralbu** [reply]
Yes.

**Reply by jrosdahl** [reply]
If it's a `.bat` file (`emcc.bat`) then I think that's why it doesn't work. Ccache uses the win32 [`CreateProcess` API to run the compiler](https://github.com/ccache/ccache/blob/a641e5ea3087752af71e60690dcf3ed3c1dbc56a/src/ccache/execute.cpp#L241-L250), and as I understand it it's not possible to execute `.bat` scripts like that. Put differently, ccache doesn't support a `.bat` script as the compiler, it needs to be a `.exe`.

**Reply by lzralbu** [reply]
I see, so that might be it.

From these [docs](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa), it seems it might be possible to run batch files:
> To run a batch file, you must start the command interpreter; set lpApplicationName to cmd.exe and set lpCommandLine to the following arguments: /c plus the name of the batch file.

But then ccache would need to detect if it was passed a batch file and make the appropriate call, which I'm not sure if it simply amounts to checking a file extension or not.
